### ISSUE TITLE
Error during compilation of CoffeeScript code

### ISSUE THREAD

**Issue by Taritsyn** [sebastienros/jint#670]
Hello!

After updating the Jint to version 3.0.0 Beta 1612, during working of the [BundleTransformer.CoffeeScript](https://github.com/Taritsyn/BundleTransformer/wiki/CoffeeScript) module an error began to occurr:

```
TypeError: scope is undefined
   at 12731:6
```

To reproduce this error, I created a [demo project](https://www.dropbox.com/s/a8hq1mg8xp1175a/TestCoffeeScriptCompilationInJint3.zip).

**Comment by lahma** [comment]
Did this work with 2.x? I quick glimpse told me that it's actually the compiler that is throwing the exception, can be quite tricky to track down.

**Comment by Taritsyn** [comment]
Yes this code worked in version 2.0.

**Comment by Taritsyn** [comment]
In the [`JavaScriptException`](https://github.com/sebastienros/jint/blob/dev/Jint/Runtime/JavaScriptException.cs) class there is a `CallStack` property that is empty in most cases. If this property always contained a call stack, it would be easier for us to find the causes of such errors.

**Comment by Taritsyn** [comment]
> ... it's actually the compiler that is throwing the exception ...

This error is not thrown by the CoffeeScript compiler, but occurs in its source code. It’s just that at some point in time the `scope` variable turns out to be equals `undefined`.

**Comment by Taritsyn** [comment]
@lahma I determined the location in the `coffeescript-combined.js` file where the error occurs:

```
6229                        o = merge(o, {
6230                            level: LEVEL_TOP
6231                        });
```
To be more precise, the error occurs in the `extend` method when copying the `scope` property:

```
153        // Extend a source object with the properties of another object (shallow copy).
154        extend = exports.extend = function (object, properties) {
155            var key, val;
156            for (key in properties) {
157                val = properties[key];
158                object[key] = val;
159            }
160            return object;
161        };
```

**Comment by Taritsyn** [comment]
The above is reason for disappearance of the `scope` property.  The “TypeError: scope is undefined” error itself occurs on line 6247 when accessing to property of the `scope` variable:

```
6243                        post = this.compileNode(o);
6244                        var _o2 = o;
6245                        scope = _o2.scope;
6246
6247                        if (scope.expressions === this) {
6248                            declars = o.scope.hasDeclarations();
```

**Comment by lahma** [comment]
@Taritsyn thank you for the details, it helped me narrow the thing down. This was caused by faulty enumerator in StringDictionarySlim that acted up after delete that hit the fix bucket index - so a nice corner(ish) case.

PR https://github.com/sebastienros/jint/pull/677 should fix this issue.

**Comment by Taritsyn** [comment]
@lahma Thank you very much!

**Comment by lahma** [comment]
@Taritsyn available on NuGet feed now 🚀

**Comment by sebastienros** [comment]
I think @Taritsyn already included this fix. I had published it on Nuget before today's fix.

**Comment by Taritsyn** [comment]
I have already released [version 3.2.2](https://github.com/Taritsyn/JavaScriptEngineSwitcher/releases/tag/v3.2.2) which supports the Jint version 3.0.0 Beta 1629.

**Comment by lahma** [comment]
Great, sorry for the confusion, hard to keep up with all the changes 👍🏻
